webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -406,43 +409,51 @@ trait DefaultFindRuleNodeStatusReports extends ReportingService {
     , unexpectedInterpretation: UnexpectedReportInterpretation
   ): Box[Map[NodeId, NodeStatusReport]] = {
 
-    val t0 = System.currentTimeMillis
-
-    /*
-     * We want to optimize and only query reports for nodes that we
-     * actually want to merge/compare or report as unexpected reports
-     */
-    val agentRunIds = (runInfos.collect { case(nodeId, run:LastRunAvailable) =>
-       AgentRunId(nodeId, run.lastRunDateTime)
-    }).toSet ++ (runInfos.collect { case(nodeId, Pending(_, Some(run), _)) =>
-       AgentRunId(nodeId, run._1)
-    }).toSet
+    var u1, u2 = 0L
 
-    for {
+    val batchedRunsInfos = runInfos.grouped(jdbcMaxBatchSize).toSeq
+    val result = sequence(batchedRunsInfos) { runBatch =>
+      val t0 = System.nanoTime()
       /*
-       * now get reports for agent rules.
-       * We want to do a batch query for all nodes to be able to minimize number of requests.
-       *
-       * We don't want to do the query if we are in "reports-disabled" mode, since in that mode,
-       * either we don't have reports (expected) or we have reports that will be out of date
-       * (for ex. just after a change in the option).
+       * We want to optimize and only query reports for nodes that we
+       * actually want to merge/compare or report as unexpected reports
        */
-      reports <- complianceModeName match {
-                                      case ReportsDisabled => Full(Map[NodeId,Seq[Reports]]())
-                                      case _               => reportsRepository.getExecutionReports(agentRunIds, ruleIds)
-                                    }
-      t1      =  System.currentTimeMillis
-      _       =  TimingDebugLogger.trace(s"Compliance: get Execution Reports for ${runInfos.size} runInfos: ${t1-t0}ms")
-    } yield {
-      val t2 = System.currentTimeMillis
-      //we want to have nodeStatus for all asked node, not only the ones with reports
-      val nodeStatusReports = runInfos.par.map { case (nodeId, runInfo) =>
-        val status = ExecutionBatch.getNodeStatusReports(nodeId, runInfo, reports.getOrElse(nodeId, Seq()), unexpectedInterpretation)
-        (status.nodeId, status)
+      val agentRunIds = (runBatch.collect { case (nodeId, run: LastRunAvailable) =>
```

Comment:
I'm not sure why we do 2 traverse of runBatch here (I know it was like that before). It seems that we could just do:
```
runBatch.flatMap { case (nodeId, run) => run match {
  case r: LastRunAvailable    => Some(AgentRunId(nodeId, run.lastRunDateTime))
  case Pending(_, Some(r), _) => Some(AgentRunId(nodeId, r._1))
  case _                      => None
} }
```

Reply:
indeed